src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,394 @@
+
+        // Combine the `innerAdapted` and `cleanup` action into a try/finally block.
+        // (Arena, C1-Cn)(int|long)
+        final MethodHandle tryFinally = MethodHandles.tryFinally(innerAdapted, cleanup);
+
+        // Acquire the arena from the global pool.
+        // With this, we finally arrive at the intended method handle:
+        // (C1-Cn)(int|long)
+        return MethodHandles.collectArguments(tryFinally, 0, HANDLES_CACHE.apply(ACQUIRE_ARENA));
+    }
+
+    private static MethodHandle makeBasicHandle(BasicKey basicKey) {
+        final VarHandle vh = CAPTURE_LAYOUT.varHandle(
+                MemoryLayout.PathElement.groupElement(basicKey.stateName()));
+        // This MH is used to extract the named captured state
+        // from the capturing `MemorySegment`.
+        // (MemorySegment, long)int
+        MethodHandle intExtractor = vh.toMethodHandle(VarHandle.AccessMode.GET);
+        // As the MH is already adapted to use the appropriate
+        // offset, we just insert `0L` for the offset.
+        // (MemorySegment)int
+        intExtractor = MethodHandles.insertArguments(intExtractor, 1, 0L);
+
+        // If X is the `returnType` (either `int` or `long`) then
+        // the code below is equivalent to:
+        //
+        // X handle(X returnValue, MemorySegment segment)
+        //     if (returnValue >= 0) {
+        //         // Ignore the segment
+        //         return returnValue;
+        //     } else {
+        //         // ignore the returnValue
+        //         return -(X)intExtractor.invokeExact(segment);
+        //     }
+        // }
+        if (basicKey.returnType().equals(int.class)) {
+            // (int, MemorySegment)int
+            return MethodHandles.guardWithTest(
+                    HANDLES_CACHE.apply(NON_NEGATIVE_INT),
+                    HANDLES_CACHE.apply(SUCCESS_INT),
+                    HANDLES_CACHE.apply(ERROR_INT).bindTo(intExtractor));
+        } else {
+            // (long, MemorySegment)long
+            return MethodHandles.guardWithTest(
+                    HANDLES_CACHE.apply(NON_NEGATIVE_LONG),
+                    HANDLES_CACHE.apply(SUCCESS_LONG),
+                    HANDLES_CACHE.apply(ERROR_LONG).bindTo(intExtractor));
+        }
+    }
+
+    // The methods below are reflective used via static MethodHandles
+
+    @ForceInline
+    private static Arena acquireArena() {
+        return POOL.pushFrame(CAPTURE_LAYOUT);
+    }
+
+    @ForceInline
+    private static MemorySegment allocate(Arena arena) {
+        // We do not need to zero out the segment.
+        return ((NoInitSegmentAllocator) arena)
+                .allocateNoInit(CAPTURE_LAYOUT.byteSize(), CAPTURE_LAYOUT.byteAlignment());
+    }
+
+    @ForceInline
+    private static boolean nonNegative(int value) {
+        return value >= 0;
+    }
+
+    @ForceInline
+    private static int success(int value,
+                               MemorySegment segment) {
```

Comment:
```suggestion
    private static long success(long value, MemorySegment segment) {
```
Newline separation is not necessary here

Reply:
I've taken the approach to always separate parameters in this class.